### ISSUE TITLE
beforeValidate not persisting on save

### ISSUE THREAD

**Issue by Klemensas** [sequelize/sequelize#8586]
## What are you doing?

```typescript
class Test extends Model {
  public id: number;
  public test: number;
  public random: number;
}
Test.init({
  _id: {
    type: DataTypes.INTEGER,
    allowNull: false,
    primaryKey: true,
    autoIncrement: true,
  },
  test: {
    type: DataTypes.FLOAT,
  },
  random: {
    type: DataTypes.FLOAT,
  },
}, { sequelize });

Test.beforeValidate((test: Test) => {
  test.test = Math.random();
  test.changed('test', true);
});

let value;
Test.sync()
  .then(() => Test.create())
  .then((test: Test) => {
    test.random = Date.now();
    value = test.test;

    return test.save();
  })
  .then((test: Test) => test.reload())
  .then((test: Test) => {
    console.log('changed', test.test, value, value !== test.test);
  });

```

## What do you expect to happen?
`beforeValidate` should persist changes both on model creation and update.

## What is actually happening?
The `test` value remains the same, calling `save` returns the modified instance but on reload we can observe that `test` field remains unchanged.

Looks like https://github.com/sequelize/sequelize/pull/5506 only fixed issues on `create` call.

```
Executing (default): INSERT INTO "Tests" ("_id","test") VALUES (DEFAULT,0.5210737053873329) RETURNING *;
Executing (default): UPDATE "Tests" SET "random"=1509715958124 WHERE "_id" = 1
```

__Dialect: postgres
__Dialect version: any
__Database version: any
__Sequelize version: master
__Tested with master branch: yes

**Comment by sushantdhiman** [comment]
This is a really extensive test for validate hook https://github.com/sequelize/sequelize/blob/0047dca9dffe9133ae231d7e0de50bca5714634a/test/integration/hooks/validate.test.js#L43

Can you submit a test case, because tests are definitely passing on master

**Comment by Klemensas** [comment]
Please read my issue again, actually running the example code I provided also helps understanding the issue.
Or are you saying that beforeValidate changes not persisting is expected behavior?

The issue won't occur on the tests you've linked for two reasons:
- The issue only occurs when field was changed only by the hook, which isn't happening on those tests.
- Those tests rely only on save which correctly returns the changed value but it's not persisted.

> Can you submit a test case, because tests are definitely passing on master

Okay, I can't seem to be able to complete npm install for sequelize on my machine. But here's a quick test showcasing the issue:
```javascript
describe('beforeValidate save', () => {
  it('should persist validation changes', () => {
    this.User.beforeValidate(user => {
      user.username = 'test' + new Date().getTime();
    })

    this.User.create({ username: 'T', mood: 'neutral' }).then((user) => {
      expect(user.mood).to.equal('neutral');
      expect(user.username).not.to.equal('T');

      user.mood = 'sad';
      return user.save();
    }).then((user) => {
      const savedUsername = user.username;
      
      return user.reload().then((reloadedUser) => {
        expect(savedUsername).to.equal(reloadedUser.username);
      });
    })
  });
});
```

I've based it on the ones you've linked, so please place it there, after the 1st beforeEach or copy the User model from there.

**Comment by sushantdhiman** [comment]
Hmm ok, this only happens when single hook is involved, odd.

**Comment by linceln** [comment]
same problem
I have my hook like this for all the models:
```
beforeValidate: function (obj) {
    let now = Date.now();
    if (obj.isNewRecord) {
        obj.createdAt = now;
        obj.updatedAt = now;
        obj.version = 0;
    } else {
        obj.updatedAt = now;
        obj.version++;
   }
}
```
but it doesn't persist on update

**Comment by corim707** [comment]
For those running into this a workaround is to modify the options.fields array to include fields modified as part of the hook.

`
beforeValidate: function( obj, options ){
 options.fields.push('YOUR_MODIFIED_FIELD'); //Suggest doing uniqueness here to ensure duplicate fields aren't added.
`

**Comment by joacub** [comment]
for not work any solution this still not working.....

**Comment by corim707** [comment]
@joacub We're using that work around in our production code base while waiting for this to be fixed.  Can you post a snippet that shows what you're running into?

**Comment by joacub** [comment]
this is teh code @corim707 

```
const handleSlugify = sources => (instance, options, next) => {
    const changed = sources.some(slugSourceField => instance.changed(slugSourceField));

    const slugValue = instance.slug;

    if (!slugValue || (changed)) {
      const slugParts = sources.map(slugSourceField => instance[slugSourceField]);

      instance.slug = formatSlug(slugParts.join(' '), { mode: 'rfc3986' });
      return next ? next(null, instance) : instance;
    }

    instance.slug = slugValue;
    return next ? next(null, instance) : instance;
  };
  Dealers.beforeCreate('handleSlugify', handleSlugify(['name']));

  Dealers.beforeUpdate('handleSlugify', handleSlugify(['name']));
```

**Comment by MatthewHerbst** [comment]
@corim707 that fixed it for us, thank you. Any idea of _why_ that fixes it?

**Comment by ajmueller** [comment]
Thank you @corim707 for the workaround.  I just ran into this and was pounding my head against my desk trying to figure out what was going on.  The instance definitely gets modified as expected when viewing changes in a debugger, so I was terribly confused.  I hope this gets resolved soon.

**Comment by ShejaEddy** [comment]
thank you @corim707 you really saved my day

**Comment by corim707** [comment]
It's been a bit since I came back to look at the issue, but here's an explanation for why it works.

The save and update methods run hooks like 'beforeUpdate' then evaluate the changed fields to re-integrate missing items from the hooks.  You can see this in sequelize@4.44.3 in model.js lines 3616 to 3626:
`              afterHookValues = _.pick(this.dataValues, _.difference(this.changed(), ignoreChanged));

              hookChanged = [];
              for (const key of Object.keys(afterHookValues)) {
                if (afterHookValues[key] !== beforeHookValues[key]) {
                  hookChanged.push(key);
                }
              }

              options.fields = _.uniq(options.fields.concat(hookChanged));
`
In the validator code the same work isn't done.  See instance-validator.js lines 103 to 113.

`  const runHooks = this.modelInstance.constructor.runHooks.bind(this.modelInstance.constructor);
    return runHooks('beforeValidate', this.modelInstance, this.options)
      .then(() =>
        this._validate()
        .catch(error => runHooks('validationFailed', this.modelInstance, this.options, error)
         .then(newError => { throw newError || error; }))
      )
      .then(() => runHooks('afterValidate', this.modelInstance, this.options))
      .return(this.modelInstance); 
`

TLDR; By injecting your own fields into the list you perform the merge that sequelize does in the case of other hooks.

**Comment by MJB-Khorasani** [comment]
> this is teh code @corim707
> 
> ```
> const handleSlugify = sources => (instance, options, next) => {
>     const changed = sources.some(slugSourceField => instance.changed(slugSourceField));
> 
>     const slugValue = instance.slug;
> 
>     if (!slugValue || (changed)) {
>       const slugParts = sources.map(slugSourceField => instance[slugSourceField]);
> 
>       instance.slug = formatSlug(slugParts.join(' '), { mode: 'rfc3986' });
>       return next ? next(null, instance) : instance;
>     }
> 
>     instance.slug = slugValue;
>     return next ? next(null, instance) : instance;
>   };
>   Dealers.beforeCreate('handleSlugify', handleSlugify(['name']));
> 
>   Dealers.beforeUpdate('handleSlugify', handleSlugify(['name']));
> ```

Hi, I'm newbie. how can i add this codes in my model. this is my model: 
const { Model, DataTypes } = require('sequelize');
const createDomPurify = require('dompurify');
const { JSDOM } = require('jsdom');
const slugify = require('slugify');
const marked = require('marked');
const dompurify = createDomPurify(new JSDOM().window);

class PostModel extends Model { };
PostModel.init({
    id: {
        primaryKey: true,
        type: DataTypes.UUID,
        defaultValue: DataTypes.UUIDV4
    },
    title: {
        type: DataTypes.STRING,
        allowNull: false,
        validate: {
            notNull: true,
            notEmpty: true
        }
    },
    description: {
        type: DataTypes.STRING,
        allowNull: false,
        validate: {
            notNull: true,
            notEmpty: true
        }
    },
    markedContent: {
        type: DataTypes.TEXT,
        allowNull: false,
        validate: {
            notNull: true,
            notEmpty: true
        }
    },
    sanitizedContent: DataTypes.TEXT,
    slug: {
        type: DataTypes.STRING,
        unique: true
    },
    thumbnail: {
        type: DataTypes.STRING,
        allowNull: false,
        // validate: {
        //     isUrl: true
        // }
    },
    baner: {
        type: DataTypes.STRING,
        allowNull: false,
        // validate: {
        //     isUrl: true
        // }
    },
    isPublished: {
        type: DataTypes.BOOLEAN,
        allowNull: true
    }
}, {
    sequelize,
    modelName: 'posts',
    timestamps: true,
    paranoid: false,
    hooks: {
        beforeValidate(post) {
            post.slug = slugify(post.title, { lower: true, remove: /[`~!@#$%^&*()|+=?;:'",.<>\{\}\[\]\\\/]/gi }); // this line
            post.sanitizedContent = dompurify.sanitize(marked(post.markedContent));                           // and this has no effect in result
        }
    }
});
module.exports = PostModel;

**Comment by MatayoshiMariano** [comment]
> beforeValidate: function( obj, options ){
>  options.fields.push('YOUR_MODIFIED_FIELD');
>  //Suggest doing uniqueness here to ensure duplicate fields aren't added.
> }

Almost did the job, but ended up doing:
```js
beforeValidate: function( obj, options ){
  obj.attrA = obj.attrB + obj.attrC
  options.fields ||= []
  options.fields.push("attrA");
  options.fields = _.uniq(options.fields);
}
```

**Comment by cirosantilli** [comment]
Here's a minimal external reproducer BTW, fails on sequelize 6.5.1 node 14.17.0 Ubuntu 21.04:

```
#!/usr/bin/env node

// https://github.com/sequelize/sequelize/issues/3534
// https://github.com/sequelize/sequelize/issues/8586

const assert = require('assert');
const path = require('path');

const { Sequelize, DataTypes } = require('sequelize');

const sequelize = new Sequelize({
  dialect: 'sqlite',
  storage: 'tmp.' + path.basename(__filename) + '.sqlite',
});

(async () => {

const IntegerNames = sequelize.define('IntegerNames',
  {
    value: {
      type: DataTypes.INTEGER,
      allowNull: false,
      unique: true,
    },
    name: {
      type: DataTypes.STRING,
    },
    name2: {
      type: DataTypes.STRING,
    },

  },
  {
    hooks: {
      beforeValidate: (integerName, options) => {
        integerName.name2 = integerName.name + 'asdf'
        //options.fields.push('name2');
      }
    ,}
  },
);
await IntegerNames.sync({force: true})
await IntegerNames.create({value: 2, name: 'two'});
await IntegerNames.create({value: 3, name: 'three'});
await IntegerNames.create({value: 5, name: 'five'});

const integerName = await IntegerNames.findOne({ where: { value: 2 } });
assert.strictEqual(integerName.name, 'two');
assert.strictEqual(integerName.name2, 'twoasdf');
integerName.name = 'TWO'
integerName.save();

const integerName2 = await IntegerNames.findOne({ where: { value: 2 } });
assert.strictEqual(integerName2.name, 'TWO');
assert.strictEqual(integerName2.name2, 'TWOasdf');

await sequelize.close();

})();
```

**Comment by Virthuss** [comment]
We are having the issue as well. Is there any fix coming? 
sequelize 1.22.5, node v12.18.4

**Comment by corim707** [comment]
> Almost did the job, but ended up doing:
> 
> ```js
> beforeValidate: function( obj, options ){
>   obj.attrA = obj.attrB + obj.attrC
>   options.fields ||= []
>   options.fields.push("attrA");
>   options.fields = _.uniq(options.fields);
> }
> ```

AFAIK, options.fields will always be an array so no need for the assignment. In pratice, the _.uniq check isn't critical but it may otherwise result in your field being validated twice.
